### DISCUSSION TITLE
Williams %R

### DISCUSSION THREAD

**Discussion by DaveSkender** [DaveSkender/Stock.Indicators#229] [Indicators]
Community discussion about [Williams %R](https://dotnet.stockindicators.dev/indicators/WilliamsR/#content)

**Comment by sshquack** [comment]
@DaveSkender I'm trying to add the Williams %R to https://charts.stockindicators.dev/ but it doesn't show up as an option. Could this be related to https://github.com/DaveSkender/Stock.Indicators/discussions/477#discussioncomment-4492623 ?

**Reply by DaveSkender** [reply]
It's not integrated yet, see https://github.com/DaveSkender/Stock.Charts/issues/212

**Reply by bsambrone** [reply]
I think there may be a calculation issue with this indicator? I don't think it should ever go below -100, but when I was testing with Agilent (A) I found this occurence many times. I never found it to go higher than 0, which is good. This is how I found the suspicious instances:

```
          var symbol = "A"; // agilent
          var loader = new Loader(db);
          List<Quote> prices = await loader.LoadPricesBySymbolAsync(symbol);

          var williams = prices.GetWilliamsR(14).ToList();
          DateTime lastBad = DateTime.MinValue;
          int below100bad = 0;
          int below80 = 0;
          int below20above80 = 0;
          int above20 = 0;
          int above0bad = 0;
          for (int i = 0; i < williams.Count; i++)
          {
              if (williams[i].WilliamsR < -100) {
                  lastBad = williams[i].Date;
                  below100bad++;
              }
              if (williams[i].WilliamsR >= -100 && williams[i].WilliamsR < -80) below80++;
              if (williams[i].WilliamsR >= -80 && williams[i].WilliamsR < -20) below20above80++;
              if (williams[i].WilliamsR >= -20 && williams[i].WilliamsR <= 0) above20++;
              if (williams[i].WilliamsR > 0) above0bad++;
          }
          
          Console.WriteLine("done, inspect yer stuff in the debugger");
```

**Reply by DaveSkender** [reply]
We analyzed this and were only able to find `100.00000000000001` out of boundary values when using unrealistic extreme price swings and significant digits, where the math exceeds the 16-digit precisions limits of the `double` data type.

I suspect the errors above are false negatives due to [conditionally expressions](https://dotnetfiddle.net/IGDK19) of `null` for initial warmup values for %R.  Maybe.  I thought what you've shown `(williams[i].WilliamsR < -100)` would evaluate as `False` when it's a `null` value, but I also see the same `null` value would fail `Assert.IsTrue(value >= -100)` in a unit test.

More info:

- #1127
- #649
